storm-core/src/jvm/org/apache/storm/daemon/supervisor/ReadClusterState.java

```diff
@@ -0,0 +1,328 @@
+                            Set<TopoProfileAction> actions = filtered.get(port);
+                            if (actions == null) {
+                                actions = new HashSet<>();
+                                filtered.put(port.intValue(), actions);
+                            }
+                            actions.add(new TopoProfileAction(topoId, req));
+                        }
+                    }
+                }
+            }
+            
+            for (Integer port: allPorts) {
+                Slot slot = slots.get(port);
+                if (slot == null) {
+                    slot = mkSlot(port);
+                    slots.put(port, slot);
+                    slot.start();
+                }
+                slot.setNewAssignment(allAssignments.get(port));
+                slot.addProfilerActions(filtered.get(port));
+            }
+            
+        } catch (Exception e) {
+            LOG.error("Failed to Sync Supervisor", e);
+            throw new RuntimeException(e);
+        }
+    }
+    
+    protected Map<String, VersionedData<Assignment>> getAssignmentsSnapshot(IStormClusterState stormClusterState, List<String> topoIds,
+            Map<String, VersionedData<Assignment>> localAssignmentVersion, Runnable callback) throws Exception {
+        Map<String, VersionedData<Assignment>> updateAssignmentVersion = new HashMap<>();
+        for (String topoId : topoIds) {
+            Integer recordedVersion = -1;
+            Integer version = stormClusterState.assignmentVersion(topoId, callback);
+            VersionedData<Assignment> locAssignment = localAssignmentVersion.get(topoId);
+            if (locAssignment != null) {
+                recordedVersion = locAssignment.getVersion();
+            }
+            if (version == null) {
+                // ignore
+            } else if (version == recordedVersion) {
+                updateAssignmentVersion.put(topoId, locAssignment);
+            } else {
+                VersionedData<Assignment> assignmentVersion = stormClusterState.assignmentInfoWithVersion(topoId, callback);
+                updateAssignmentVersion.put(topoId, assignmentVersion);
+            }
+        }
+        return updateAssignmentVersion;
+    }
+    
+    protected Map<String, List<ProfileRequest>> getProfileActions(IStormClusterState stormClusterState, List<String> stormIds) throws Exception {
+        Map<String, List<ProfileRequest>> ret = new HashMap<String, List<ProfileRequest>>();
+        for (String stormId : stormIds) {
+            List<ProfileRequest> profileRequests = stormClusterState.getTopologyProfileRequests(stormId);
+            ret.put(stormId, profileRequests);
+        }
+        return ret;
+    }
+    
+    protected Map<Integer, LocalAssignment> readAssignments(Map<String, VersionedData<Assignment>> assignmentsSnapshot,
+            String assignmentId, AtomicInteger retries) {
```

Comment:
This came from the original clojure code, that was then translated into java, and I just left it.  The way it works right now is that there is a sleep in between the retries in the main thread.  I wanted to fix it, but felt leaving it would be simpler.  If you want me to I will change it.